website/oauth/models/__init__.py

```diff
@@ -0,0 +1,325 @@
+                verifier=request.args.get('oauth_verifier'),
+            ).fetch_access_token(self.callback_url)
+
+        elif self._oauth_version == OAUTH2:
+            state = request.args.get('state')
+
+            if cached_credentials.get('state') != state:
+                raise PermissionsError("Request token does not match")
+
+            response = OAuth2Session(
+                self.client_id,
+                redirect_uri=web_url_for(
+                    'oauth_callback',
+                    service_name=self.short_name,
+                    _absolute=True
+                ),
+            ).fetch_token(
+                self.callback_url,
+                client_secret=self.client_secret,
+                code=request.args.get('code'),
+            )
+
+        info = self._default_handle_callback(response)
+        info.update(self.handle_callback(response))
+
+        try:
+            self.account = ExternalAccount.find_one(
+                Q('provider', 'eq', self.short_name) &
+                Q('provider_id', 'eq', info['provider_id'])
+            )
+        except NoResultsFound:
+            self.account = ExternalAccount(provider=self.short_name,
+                                           provider_id=info['provider_id'])
+
+        # required
+        self.account.oauth_key = info['key']
+
+        # only for OAuth1
+        self.account.oauth_secret = info.get('secret')
+
+        # only for OAuth2
+        self.account.expires_at = info.get('expires_at')
+        self.account.refresh_token = info.get('refresh_token')
+
+        # additional information
+        self.account.display_name = info.get('display_name')
+
+        self.account.save()
+
+        if self.account not in user.external_accounts:
+            user.external_accounts.append(self.account)
+            user.save()
+
+    def _default_handle_callback(self, data):
+        if self._oauth_version == OAUTH1:
+            key = data.get('oauth_token')
+            secret = data.get('oauth_token_secret')
+
+            values = {}
+
+            if key:
+                values['key'] = key
+            if secret:
+                values['secret'] = secret
+
+            return values
+
+        elif self._oauth_version == OAUTH2:
+            key = data.get('access_token')
+            refresh_token = data.get('refresh_token')
+            expires_at = data.get('expires_at')
+            scopes = data.get('scope')
+
+            values = {}
+
+            if key:
+                values['key'] = key
+            if scopes:
+                values['scope'] = scopes
+            if refresh_token:
+                values['refresh_token'] = refresh_token
+            if expires_at:
+                values['expires_at'] = datetime.datetime.fromtimestamp(
+                    float(expires_at)
+                )
+
+            return values
+
+    def handle_callback(self, response):
+        return {}
+
+
+class Zotero(ExternalProvider):
```

Comment:
Could these providers be moved into their own modules? Also, can we delete the providers that aren't being used to avoid confusion?